pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderTokenJWK.java

```
@@ -0,0 +1,363 @@
+            throws AuthenticationException {
+        return new TokenAuthenticationState(this, authData, remoteAddress, sslSession);
+    }
+
+    public static String getToken(AuthenticationDataSource authData) throws AuthenticationException {
+        if (authData.hasDataFromCommand()) {
+            // Authenticate Pulsar binary connection
+            return validateToken(authData.getCommandData());
+        } else if (authData.hasDataFromHttp()) {
+            // Authentication HTTP request. The format here should be compliant to RFC-6750
+            // (https://tools.ietf.org/html/rfc6750#section-2.1). Eg: Authorization: Bearer xxxxxxxxxxxxx
+            String httpHeaderValue = authData.getHttpHeader(HTTP_HEADER_NAME);
+            if (httpHeaderValue == null || !httpHeaderValue.startsWith(HTTP_HEADER_VALUE_PREFIX)) {
+                throw new AuthenticationException("Invalid HTTP Authorization header");
+            }
+
+            // Remove prefix
+            String token = httpHeaderValue.substring(HTTP_HEADER_VALUE_PREFIX.length());
+            return validateToken(token);
+        } else {
+            throw new AuthenticationException("No token credentials passed");
+        }
+    }
+
+    private static String validateToken(final String token) throws AuthenticationException {
+        if (StringUtils.isNotBlank(token)) {
+            return token;
+        } else {
+            throw new AuthenticationException("Blank token found");
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private DecodedJWT authenticateToken(final String token) throws AuthenticationException {
+
+        DecodedJWT jwt = JWT.decode(token);
+
+        Jwk jwk = null;
+        Algorithm algorithm = null;
+        try {
+            jwk = provider.get(jwt.getKeyId());
+
+            if(!(jwk instanceof RSAPublicKey)){
+                throw new JwtException("key needs to be a RSA Publickey");
+            }
```

@michaeljmarshall I'm expecting the key to be an instance of RSAPublicKey so I want to throw an exception if it's not. I'm not sure if you had something else in mind 😄